src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithFILTER.java

```diff
@@ -0,0 +1,82 @@
+package se.liu.ida.hefquin.engine.queryplan.executable.impl.ops;
+
+import java.util.Iterator;
+import java.util.Set;
+
+import org.apache.jena.graph.Node;
+import org.apache.jena.sparql.algebra.Op;
+import org.apache.jena.sparql.algebra.op.OpFilter;
+import org.apache.jena.sparql.algebra.op.OpTriple;
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import org.apache.jena.sparql.expr.E_Equals;
+import org.apache.jena.sparql.expr.E_LogicalAnd;
+import org.apache.jena.sparql.expr.E_LogicalOr;
+import org.apache.jena.sparql.expr.Expr;
+import org.apache.jena.sparql.expr.ExprVar;
+import org.apache.jena.sparql.expr.nodevalue.NodeValueNode;
+
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
+import se.liu.ida.hefquin.engine.federation.access.SolMapsResponse;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.SPARQLRequestImpl;
+import se.liu.ida.hefquin.engine.query.SPARQLGraphPattern;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+import se.liu.ida.hefquin.engine.query.impl.QueryPatternUtils;
+import se.liu.ida.hefquin.engine.query.impl.SPARQLGraphPatternImpl;
+import se.liu.ida.hefquin.engine.queryproc.ExecutionContext;
+
+public class ExecOpBindJoinSPARQLwithFILTER extends ExecOpGenericBindJoin<TriplePattern,SPARQLEndpoint> {
+	
+	protected final Set<Var> varsInTP;
+
+	public ExecOpBindJoinSPARQLwithFILTER(TriplePattern query, SPARQLEndpoint fm) {
+		super(query, fm);
+		varsInTP = QueryPatternUtils.getVariablesInPattern(query);
+	}
+
+	@Override
+	protected Iterable<? extends SolutionMapping> fetchSolutionMappings(Iterable<SolutionMapping> solMaps,
+			ExecutionContext execCxt) {
+		final Op op = createFilter(solMaps);
+		final SPARQLGraphPattern pattern = new SPARQLGraphPatternImpl(op);
+		final SPARQLRequest request = new SPARQLRequestImpl(pattern);
+		final SolMapsResponse response = execCxt.getFederationAccessMgr().performRequest(request, fm);
+		return response.getSolutionMappings();
+	}
+
+	private Op createFilter(Iterable<SolutionMapping> solMaps) {
+		Expr disjunction = null;
+		boolean firstDisj = true;
+		for (final SolutionMapping s : solMaps) {
+			Binding b = SolutionMappingUtils.restrict(s, varsInTP).asJenaBinding();
```

Comment:
This can be done more efficiently: the restrict method that you use here creates a SolutionMapping object, which is not actually needed here. Therefore, please add another restrict method to SolutionMappingUtils that takes a Binding as argument (rather than a SolutionMapping) and that returns a Binding. Then, use that method here. Additionally, you can then use that method also in the implementation of the existing restrict method.